In [1]:
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np

In [2]:
cs_all_emh = pd.read_excel('./cs_all_emh.xlsx', index_col =0)
corr_20012020 = pd.read_excel('./corr_20012020.xlsx', usecols ='B:E')

In [5]:
cs_regre = pd.concat([cs_all_emh, corr_20012020], axis =1)

In [6]:
cs_regre.rename(columns = {'year':'년도','elementary':'초등학교','middle':'중학교','high':'고등학교','all':'폐교',
                          'Elementary':'초등학생','Middle':'중학생','High':'고등학생','All':'학령인구'},inplace=True)

In [54]:
cs_regre.head()

년도  초등학교  중학교  고등학교  폐교  초등학생   중학생  고등학생  학령인구
0  2001    24    2     2  28  4150  1847  2014  8011
1  2002    23    3     0  26  4192  1863  1902  7957
2  2003     9    0     1  10  4186  1913  1846  7945
3  2004    13    2     1  16  4118  1993  1824  7935
4  2005    16    4     1  21  4018  2064  1841  7923

# 회귀분석

In [8]:
import numpy as np
import matplotlib.pyplot as plt
from scipy.optimize import curve_fit
import statsmodels.api as sm
import statsmodels.formula.api as smf
from statsmodels.formula.api import ols

In [45]:
#초등학생
model = smf.ols(formula = '폐교 ~ 초등학생', data =cs_regre)
result_ele = model.fit()
result_ele.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                     폐교   R-squared:                       0.298
Model:                            OLS   Adj. R-squared:                  0.259
Method:                 Least Squares   F-statistic:                     7.626
Date:                Sat, 11 Dec 2021   Prob (F-statistic):             0.0129
Time:                        19:17:36   Log-Likelihood:                -70.210
No. Observations:                  20   AIC:                             144.4
Df Residuals:                      18   BIC:                             146.4
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     59.3569     10.961      5.415      0.000      36.328      82.386
초등학생          -0.0089      0.003     -2.761      0.013      -0.016      -0.002
==============================================================================
Omnibus:                        1.537   Durbin-Watson:                   1.655
Prob(Omnibus):                  0.464   Jarque-Bera (JB):                0.994
Skew:                           0.205   Prob(JB):                        0.608
Kurtosis:                       1.987   Cond. No.                     1.94e+04
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 1.94e+04. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

- R-squared:	0.298  => 회귀모형의 설명력이 29.8%라는 뜻
- Intercept(절편)의 t값은 5.415
- P>|t|값은 0.000으로 이 회귀모형의 절편값은 유의한 것으로 나타남/ 값은 59.3569

- 초등학생의 t = -2.761, P>|t|값은 0.013, 초등학생의 계수는 -0.0089로 도출

### 회귀모형의 식은 Y = 59.3569 + (-0.0089)*초등학생 

In [51]:
#초등학생 / 초등학교
model = smf.ols(formula = '초등학교 ~ 초등학생', data =cs_regre)
result_ele = model.fit()
result_ele.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                   초등학교   R-squared:                       0.026
Model:                            OLS   Adj. R-squared:                 -0.028
Method:                 Least Squares   F-statistic:                    0.4874
Date:                Sat, 11 Dec 2021   Prob (F-statistic):              0.494
Time:                        20:06:17   Log-Likelihood:                -68.262
No. Observations:                  20   AIC:                             140.5
Df Residuals:                      18   BIC:                             142.5
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     28.6864      9.944      2.885      0.010       7.794      49.578
초등학생          -0.0021      0.003     -0.698      0.494      -0.008       0.004
==============================================================================
Omnibus:                        1.126   Durbin-Watson:                   1.640
Prob(Omnibus):                  0.569   Jarque-Bera (JB):                0.815
Skew:                           0.472   Prob(JB):                        0.665
Kurtosis:                       2.704   Cond. No.                     1.94e+04
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 1.94e+04. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

- R-squared:0.026  => 회귀모형의 설명력이 2.6%라는 뜻
- Intercept(절편)의 t값은 2.885
- P>|t|값은 0.010으로 이 회귀모형의 절편값은 유의한 것으로 나타남/ 값은 28.6864

- 초등학생의 t = -0.698, P>|t|값은 0.494, 초등학생의 계수는 -0.0021로 도출

### 회귀모형의 식은 Y = 28.6864 + (-0.0021)*초등학생 

In [10]:
#중학생

model = smf.ols(formula = '폐교 ~ 중학생', data =cs_regre)
result = model.fit()
result.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                     폐교   R-squared:                       0.163
Model:                            OLS   Adj. R-squared:                  0.117
Method:                 Least Squares   F-statistic:                     3.510
Date:                Sat, 11 Dec 2021   Prob (F-statistic):             0.0773
Time:                        19:06:25   Log-Likelihood:                -71.961
No. Observations:                  20   AIC:                             147.9
Df Residuals:                      18   BIC:                             149.9
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     55.3026     13.903      3.978      0.001      26.094      84.511
중학생           -0.0144      0.008     -1.873      0.077      -0.031       0.002
==============================================================================
Omnibus:                        0.041   Durbin-Watson:                   1.377
Prob(Omnibus):                  0.980   Jarque-Bera (JB):                0.263
Skew:                           0.018   Prob(JB):                        0.877
Kurtosis:                       2.440   Cond. No.                     1.21e+04
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 1.21e+04. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

- R-squared:	0.163  => 회귀모형의 설명력이 16.3%라는 뜻
- Intercept(절편)의 t값은 3.978
- P>|t|값은 0.001으로 이 회귀모형의 절편값은 유의한 것으로 나타남/ 값은 55.3026

- 중학생의 t = -1.873, P>|t|값은 0.077, 중학생의 계수는 -0.0144로 도출

### 회귀모형의 식은 Y = 55.3026 + (-0.0144)*중학생 

In [52]:
#중학생 / 중학교

model = smf.ols(formula = '중학교 ~ 중학생', data =cs_regre)
result = model.fit()
result.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                    중학교   R-squared:                       0.614
Model:                            OLS   Adj. R-squared:                  0.592
Method:                 Least Squares   F-statistic:                     28.62
Date:                Sat, 11 Dec 2021   Prob (F-statistic):           4.38e-05
Time:                        20:35:06   Log-Likelihood:                -53.720
No. Observations:                  20   AIC:                             111.4
Df Residuals:                      18   BIC:                             113.4
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     35.4897      5.585      6.355      0.000      23.756      47.223
중학생           -0.0165      0.003     -5.350      0.000      -0.023      -0.010
==============================================================================
Omnibus:                        1.586   Durbin-Watson:                   1.165
Prob(Omnibus):                  0.453   Jarque-Bera (JB):                0.875
Skew:                           0.512   Prob(JB):                        0.646
Kurtosis:                       2.969   Cond. No.                     1.21e+04
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 1.21e+04. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

- R-squared:	0.614  => 회귀모형의 설명력이 61.4%라는 뜻
- Intercept(절편)의 t값은 6.355
- P>|t|값은 0.000으로 이 회귀모형의 절편값은 유의한 것으로 나타남/ 값은 35.4897

- 중학생의 t = -5.350, P>|t|값은 0.000, 중학생의 계수는 -0.0165로 도출

### 회귀모형의 식은 Y = 35.4897 + (-0.0165)*중학생 

In [11]:
#고등학생
model = smf.ols(formula = '폐교 ~ 고등학생', data =cs_regre)
result = model.fit()
result.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                     폐교   R-squared:                       0.011
Model:                            OLS   Adj. R-squared:                 -0.044
Method:                 Least Squares   F-statistic:                    0.1920
Date:                Sat, 11 Dec 2021   Prob (F-statistic):              0.666
Time:                        19:06:27   Log-Likelihood:                -73.636
No. Observations:                  20   AIC:                             151.3
Df Residuals:                      18   BIC:                             153.3
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     39.0500     21.796      1.792      0.090      -6.742      84.842
고등학생          -0.0051      0.012     -0.438      0.666      -0.030       0.019
==============================================================================
Omnibus:                        0.886   Durbin-Watson:                   1.197
Prob(Omnibus):                  0.642   Jarque-Bera (JB):                0.789
Skew:                          -0.426   Prob(JB):                        0.674
Kurtosis:                       2.529   Cond. No.                     1.80e+04
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 1.8e+04. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

- R-squared:	0.011  => 회귀모형의 설명력이 1.1%라는 뜻
- Intercept(절편)의 t값은 1.792
- P>|t|값은 0.001으로 이 회귀모형의 절편값은 유의한 것으로 나타남/ 값은 39.0500

- 고등학생의 t = -0.438, P>|t|값은 0.666, 고등학생의 계수는 -0.0051로 도출

### 회귀모형의 식은 Y = 39.0500 + (-0.0051)*고등학생 

In [53]:
#고등학생 / 고등학교
model = smf.ols(formula = '고등학교 ~ 고등학생', data =cs_regre)
result = model.fit()
result.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                   고등학교   R-squared:                       0.024
Model:                            OLS   Adj. R-squared:                 -0.030
Method:                 Least Squares   F-statistic:                    0.4378
Date:                Sat, 11 Dec 2021   Prob (F-statistic):              0.517
Time:                        20:37:40   Log-Likelihood:                -36.841
No. Observations:                  20   AIC:                             77.68
Df Residuals:                      18   BIC:                             79.67
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept      4.0286      3.462      1.164      0.260      -3.246      11.303
고등학생          -0.0012      0.002     -0.662      0.517      -0.005       0.003
==============================================================================
Omnibus:                        2.350   Durbin-Watson:                   1.817
Prob(Omnibus):                  0.309   Jarque-Bera (JB):                1.764
Skew:                           0.709   Prob(JB):                        0.414
Kurtosis:                       2.676   Cond. No.                     1.80e+04
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 1.8e+04. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

- R-squared:	0.024  => 회귀모형의 설명력이 2.4%라는 뜻
- Intercept(절편)의 t값은 1.164
- P>|t|값은 0.001으로 이 회귀모형의 절편값은 유의한 것으로 나타남/ 값은 4.0286

- 고등학생의 t = -0.662, P>|t|값은 0.517, 고등학생의 계수는 -0.0012로 도출

### 회귀모형의 식은 Y = 4.0286 + (-0.0012)*고등학생 

In [12]:
#전체 연령
model = smf.ols(formula = '폐교 ~ 학령인구', data =cs_regre)
result = model.fit()
result.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                     폐교   R-squared:                       0.237
Model:                            OLS   Adj. R-squared:                  0.194
Method:                 Least Squares   F-statistic:                     5.584
Date:                Sat, 11 Dec 2021   Prob (F-statistic):             0.0296
Time:                        19:06:28   Log-Likelihood:                -71.040
No. Observations:                  20   AIC:                             146.1
Df Residuals:                      18   BIC:                             148.1
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     64.9822     15.125      4.296      0.000      33.205      96.759
학령인구          -0.0051      0.002     -2.363      0.030      -0.010      -0.001
==============================================================================
Omnibus:                        0.390   Durbin-Watson:                   1.539
Prob(Omnibus):                  0.823   Jarque-Bera (JB):                0.526
Skew:                           0.147   Prob(JB):                        0.769
Kurtosis:                       2.262   Cond. No.                     5.36e+04
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 5.36e+04. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

- R-squared:	0.237  => 회귀모형의 설명력이 23.7%라는 뜻
- Intercept(절편)의 t값은 4.296
- P>|t|값은 0.000으로 이 회귀모형의 절편값은 유의한 것으로 나타남/ 값은 64.9822

- 학령인구의 t = -2.363, P>|t|값은 0.030, 학령인구의 계수는 -0.0051로 도출

### 회귀모형의 식은 Y = 64.9822 + (-0.0051)*학령인구 

In [13]:
#전체연령 / 다중회귀
model = smf.ols(formula = '폐교 ~ 초등학생 + 중학생 + 고등학생', data = cs_regre)
result = model.fit()
result.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                     폐교   R-squared:                       0.331
Model:                            OLS   Adj. R-squared:                  0.206
Method:                 Least Squares   F-statistic:                     2.640
Date:                Sat, 11 Dec 2021   Prob (F-statistic):             0.0849
Time:                        19:06:30   Log-Likelihood:                -69.721
No. Observations:                  20   AIC:                             147.4
Df Residuals:                      16   BIC:                             151.4
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     43.8414     22.055      1.988      0.064      -2.913      90.595
초등학생          -0.0063      0.006     -1.103      0.287      -0.018       0.006
중학생           -0.0138      0.019     -0.744      0.467      -0.053       0.026
고등학생           0.0168      0.019      0.895      0.384      -0.023       0.057
==============================================================================
Omnibus:                        1.147   Durbin-Watson:                   1.650
Prob(Omnibus):                  0.563   Jarque-Bera (JB):                0.972
Skew:                           0.326   Prob(JB):                        0.615
Kurtosis:                       2.140   Cond. No.                     4.76e+04
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 4.76e+04. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

## 회귀모형식 Y = 43.8414 + (-0.0063)*초등학생 +(-0.0138)*중학생 +(0.0168)*고등학생 